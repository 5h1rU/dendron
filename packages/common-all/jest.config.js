const jestConfig = require("../../jest.config");

console.log(jestConfig);
module.exports = {
  ...jestConfig
};
