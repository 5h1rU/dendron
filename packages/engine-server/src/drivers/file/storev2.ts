import {
  DendronError,
  DEngineInitPayloadV2,
  DNodePropsV2,
  DNodeUtilsV2,
  DStoreV2,
  EngineUpdateNodesOptsV2,
  EngineWriteOptsV2,
  ENGINE_ERROR_CODES,
  NoteOptsV2,
  NotePropsDictV2,
  NotePropsV2,
  NoteUtilsV2,
  SchemaModuleDictV2,
  SchemaModulePropsV2,
  SchemaModuleV2,
  SchemaOptsV2,
  SchemaPropsDictV2,
  SchemaPropsV2,
  SchemaUtils,
  SchemaUtilsV2,
} from "@dendronhq/common-all";
import {
  DLogger,
  file2Note,
  getAllFiles,
  globMatch,
} from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import { FileMeta } from "packages/engine-server/lib";
import path from "path";
import YAML from "yamljs";

type FileMetaV2 = {
  // file name: eg. foo.md, name = foo
  prefix: string;
  // fpath: full path, eg: foo.md, fpath: foo.md
  fpath: string;
};
type FileMetaDictV2 = { [key: string]: FileMetaV2[] };

function getFileMetaV2(fpaths: string[]): FileMetaDictV2 {
  const metaDict: FileMetaDictV2 = {};
  _.forEach(fpaths, (fpath) => {
    const { name } = path.parse(fpath);
    const lvl = name.split(".").length;
    if (!_.has(metaDict, lvl)) {
      metaDict[lvl] = [];
    }
    metaDict[lvl].push({ prefix: name, fpath });
  });
  return metaDict;
}

export class ParserBaseV2 {
  constructor(public opts: { store: DStoreV2; logger: DLogger }) {}

  get logger() {
    return this.opts.logger;
  }
}

export class NoteParserV2 extends ParserBaseV2 {
  public cache: NotePropsCacheV2;

  constructor(
    public opts: { store: DStoreV2; cache: NotePropsCacheV2; logger: DLogger }
  ) {
    super(opts);
    this.cache = opts.cache;
  }

  parseFile(fpath: string[]): NotePropsV2[] {
    const fileMetaDict: FileMetaDictV2 = getFileMetaV2(fpath);
    const maxLvl = _.max(_.keys(fileMetaDict).map((e) => _.toInteger(e))) || 2;
    const notesByFname: NotePropsDictV2 = {};

    // get root note
    if (_.isUndefined(fileMetaDict[1])) {
      throw new DendronError({ status: ENGINE_ERROR_CODES.NO_ROOT_NOTE_FOUND });
    }
    const rootFile = fileMetaDict[1].find(
      (n) => n.fpath === "root.md"
    ) as FileMetaV2;
    if (!rootFile) {
      throw new DendronError({ status: ENGINE_ERROR_CODES.NO_ROOT_NOTE_FOUND });
    }
    const rootNote = this.parseNoteProps({
      fileMeta: rootFile,
      addParent: false,
    })[0];

    notesByFname[rootNote.fname] = rootNote;

    // get root of hiearchies
    let lvl = 2;
    let prevNodes: NotePropsV2[] = fileMetaDict[1]
      // don't count root node
      .filter((n) => n.fpath !== "root.md")
      .flatMap((ent) => {
        const notes = this.parseNoteProps({ fileMeta: ent, addParent: false });
        return notes;
      });
    prevNodes.forEach((ent) => {
      DNodeUtilsV2.addChild(rootNote, ent);
      notesByFname[ent.fname] = ent;
    });

    // get everything else
    while (lvl <= maxLvl) {
      const currNodes: NotePropsV2[] = (fileMetaDict[lvl] || [])
        .filter((ent) => {
          return !globMatch(["root.*"], ent.fpath);
        })
        .flatMap((ent) => {
          const node = this.parseNoteProps({
            fileMeta: ent,
            parents: prevNodes,
            notesByFname,
            addParent: true,
          });
          return node;
        });
      lvl += 1;
      currNodes.forEach((ent) => {
        notesByFname[ent.fname] = ent;
      });
      prevNodes = currNodes;
    }

    return _.values(notesByFname);
  }

  parseNoteProps(opts: {
    fileMeta: FileMetaV2;
    notesByFname?: NotePropsDictV2;
    parents?: NotePropsV2[];
    addParent: boolean;
    createStubs?: boolean;
  }): NotePropsV2[] {
    const cleanOpts = _.defaults(opts, {
      addParent: true,
      createStubs: true,
      notesByFname: {},
    });
    const ctx = "parseNoteProps";
    const root = this.opts.store.vaults[0];
    const { fileMeta, parents, notesByFname } = cleanOpts;
    let out: NotePropsV2[] = [];
    let noteProps: NotePropsV2;
    try {
      noteProps = file2Note(path.join(root, fileMeta.fpath));
    } catch (_err) {
      const err = {
        status: ENGINE_ERROR_CODES.BAD_PARSE_FOR_NOTE,
        msg: JSON.stringify({
          fname: fileMeta.fpath,
          error: _err.message,
        }),
      };
      this.logger.error({ ctx, fileMeta, err });
      throw new DendronError(err);
    }

    if (cleanOpts.addParent) {
      const parentPath = DNodeUtilsV2.dirName(fileMeta.prefix);
      let parent = _.find(parents, (p) => p.fname === parentPath) || null;

      if (!parent && !cleanOpts.createStubs) {
        const err = {
          status: ENGINE_ERROR_CODES.NO_PARENT_FOR_NOTE,
          msg: JSON.stringify({
            fname: fileMeta.fpath,
          }),
        };
        this.logger.error({ ctx, fileMeta, err });
        throw new DendronError(err);
      }
      if (!parent) {
        parent = DNodeUtilsV2.findClosestParent(
          noteProps.fname,
          notesByFname
        ) as NotePropsV2;
        const stubNodes = NoteUtilsV2.createStubs(parent, noteProps);
        stubNodes.forEach((ent2) => {
          out.push(ent2);
        });
      }
      DNodeUtilsV2.addChild(parent, noteProps);
    }
    out.push(noteProps);
    return out;
  }

  parse(fpaths: string[]): NotePropsV2[] {
    return this.parseFile(fpaths);
  }
}

export class SchemaParserV2 extends ParserBaseV2 {
  static parseFile(fpath: string, root: string): SchemaModuleV2 {
    const fname = path.parse(fpath).name;
    const schemaOpts: any = YAML.parse(
      fs.readFileSync(path.join(root, fpath), "utf8")
    );
    const version = _.isArray(schemaOpts) ? 0 : 1;
    if (version > 0) {
      return SchemaParserV2.parseSchemaModuleProps(
        schemaOpts as SchemaModulePropsV2,
        {
          fname,
          root,
        }
      );
    } else {
      const schemaDict: SchemaPropsDictV2 = {};
      (schemaOpts as SchemaOptsV2[]).map((ent) => {
        const schema = SchemaUtilsV2.create(ent);
        schemaDict[schema.id] = schema;
      });
      const maybeRoot = schemaDict["root"] as SchemaPropsV2;
      return {
        root: maybeRoot,
        schemas: schemaDict,
        fname: maybeRoot.fname,
      };
    }
  }

  static parseSchemaModuleProps(
    schemaModuleProps: SchemaModulePropsV2,
    opts: { fname: string; root: string }
  ): SchemaModuleV2 {
    const { imports, schemas } = schemaModuleProps;
    const { fname, root } = opts;
    let schemaModulesFromImport = _.flatMap(imports, (ent) => {
      return SchemaParserV2.parseFile(`${ent}.yml`, root);
    });
    const schemaPropsFromImport = schemaModulesFromImport.flatMap((mod) => {
      return _.values(mod.schemas).map((ent) => {
        const domain = SchemaUtils.fname(ent.fname);
        ent.data.pattern = ent.data.pattern || ent.id;
        ent.id = `${domain}.${ent.id}`;
        ent.fname = fname;
        ent.parent = null;
        ent.children = ent.children.map((ent) => `${domain}.${ent}`);
        return ent;
      });
    });
    const schemaPropsFromFile = schemas.map((ent) => {
      return SchemaUtilsV2.create(ent);
    });
    const schemasAll = schemaPropsFromImport.concat(schemaPropsFromFile);

    const schemasDict: SchemaPropsDictV2 = {};
    schemasAll.forEach((ent) => {
      schemasDict[ent.id] = ent;
    });
    const rootModule = SchemaUtilsV2.getModuleRoot(schemaModuleProps);
    return {
      root: rootModule,
      schemas: schemasDict,
      fname: rootModule.fname,
    };
  }

  parse(fpaths: string[], root: string): SchemaModuleV2[] {
    return fpaths.flatMap((fpath) => {
      return SchemaParserV2.parseFile(fpath, root);
    });
  }
}

type NotePropsCacheV2 = {};

type NoteEntryV2 = {
  mtime: number;
  size: number;
  hash: number;
};

type MetaEntryV2 = {
  links: any[];
  embeds: any[];
  tags: any[];
  headings: any[];
};

export class FileStorageV2 implements DStoreV2 {
  public vaults: string[];
  public notes: NotePropsDictV2;
  public schemas: SchemaModuleDictV2;
  public notesCache: NotePropsCacheV2;
  public logger: DLogger;

  constructor(props: { vaults: string[]; logger: DLogger }) {
    const { vaults, logger } = props;
    this.vaults = vaults;
    this.notes = {};
    this.schemas = {};
    this.notesCache = {};
    this.logger = logger;
    const ctx = "FileStorageV2";
    this.logger.info({ ctx, vaults });
  }

  async init(): Promise<DEngineInitPayloadV2> {
    try {
      const _schemas = await this.initSchema();
      const _notes = await this.initNotes();
      _schemas.map((ent) => {
        this.schemas[ent.root.id] = ent;
      });
      _notes.map((ent) => {
        this.notes[ent.id] = ent;
      });
      const { notes, schemas } = this;
      return { notes, schemas };
    } catch (err) {
      this.logger.error(err);
      throw err;
    }
  }

  loadNotesCache(): NotePropsCacheV2 {
    return {};
  }

  async initSchema(): Promise<SchemaModuleV2[]> {
    const ctx = "initSchema";
    this.logger.info({ ctx, msg: "enter" });
    const schemaFiles = getAllFiles({
      root: this.vaults[0],
      include: ["*.schema.yml"],
    }) as string[];
    this.logger.info({ ctx, schemaFiles });
    const root = this.vaults[0];
    if (_.isEmpty(schemaFiles)) {
      throw new DendronError({ status: ENGINE_ERROR_CODES.NO_SCHEMA_FOUND });
    }
    return new SchemaParserV2({ store: this, logger: this.logger }).parse(
      schemaFiles,
      root
    );
  }

  async initNotes(): Promise<NotePropsV2[]> {
    const ctx = "initNotes";
    this.logger.info({ ctx, msg: "enter" });
    const noteFiles = getAllFiles({
      root: this.vaults[0],
      include: ["*.md"],
    }) as string[];
    const cache = this.loadNotesCache();
    const root = this.vaults[0];
    return new NoteParserV2({ store: this, cache, logger: this.logger }).parse(
      noteFiles
    );
  }

  async updateNote(
    note: NotePropsV2,
    opts?: EngineUpdateNodesOptsV2
  ): Promise<void> {
    throw Error("not implemented");
    return;
  }

  async updateSchema(schemaModule: SchemaModulePropsV2) {
    const vault = this.vaults[0];
    const schemas = SchemaParserV2.parseSchemaModuleProps(schemaModule, {
      fname: schemaModule.schemas[0].fname,
      root: vault,
    });
    this.schemas[schemas.root.id] = schemas;
    return;
  }

  async writeNote(note: NotePropsV2, opts?: EngineWriteOptsV2): Promise<void> {
    throw Error("to implement");
    return;
  }

  async writeSchema(schemaModule: SchemaModulePropsV2) {
    return this.updateSchema(schemaModule);
  }
}
