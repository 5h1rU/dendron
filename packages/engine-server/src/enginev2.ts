import {
  ConfigWriteOpts,
  DendronConfig,
  DendronError,
  DEngineClientV2,
  DEngineDeleteSchemaRespV2,
  DEngineInitRespV2,
  DEngineMode,
  DEngineV2,
  DLink,
  DNodeTypeV2,
  DStoreV2,
  DVault,
  EngineDeleteOptsV2,
  EngineUpdateNodesOptsV2,
  EngineWriteOptsV2,
  GetNoteOptsV2,
  GetNotePayloadV2,
  NoteChangeEntry,
  NotePropsDictV2,
  NotePropsV2,
  NoteUtilsV2,
  QueryNotesOpts,
  RenameNoteOptsV2,
  RenameNotePayload,
  RespV2,
  SchemaModuleDictV2,
  SchemaModulePropsV2,
  SchemaQueryResp,
  WorkspaceOpts,
  WriteNoteResp,
} from "@dendronhq/common-all";
import {
  createLogger,
  DLogger,
  readYAML,
  VaultUtils,
  writeYAML,
} from "@dendronhq/common-server";
import _ from "lodash";
import { DConfig } from "./config";
import { FileStorageV2 } from "./drivers/file/storev2";
import { FuseEngine } from "./fuseEngine";

type DendronEngineOptsV2 = {
  wsRoot: string;
  vaultsv3: DVault[];
  forceNew?: boolean;
  store?: any;
  mode?: DEngineMode;
  logger?: DLogger;
};
type DendronEnginePropsV2 = Required<DendronEngineOptsV2>;

export class DendronEngineV2 implements DEngineV2 {
  public wsRoot: string;
  public vaults: string[];
  public store: DStoreV2;
  protected props: DendronEnginePropsV2;
  public logger: DLogger;
  public fuseEngine: FuseEngine;
  public links: DLink[];
  public vaultsv3: DVault[];
  public configRoot: string;

  static _instance: DendronEngineV2 | undefined;

  constructor(props: DendronEnginePropsV2) {
    this.wsRoot = props.wsRoot;
    this.configRoot = props.wsRoot;
    this.store = props.store;
    this.logger = props.logger;
    this.props = props;
    this.fuseEngine = new FuseEngine({});
    this.links = [];
    this.vaultsv3 = props.vaultsv3;
    this.vaults = props.vaultsv3.map((ent) => ent.fsPath);
  }

  static createV3({
    vaults,
    wsRoot,
    logger,
  }: WorkspaceOpts & { logger?: DLogger }) {
    const LOGGER = logger || createLogger();
    return new DendronEngineV2({
      wsRoot,
      vaultsv3: vaults,
      forceNew: true,
      store: new FileStorageV2({
        wsRoot,
        vaultsv3: vaults,
        logger: LOGGER,
      }),
      mode: "fuzzy",
      logger: LOGGER,
    });
  }

  static instance({ vaults, wsRoot }: { vaults: DVault[]; wsRoot: string }) {
    if (!DendronEngineV2._instance) {
      DendronEngineV2._instance = DendronEngineV2.createV3({ vaults, wsRoot });
    }
    return DendronEngineV2._instance;
  }

  get notes(): NotePropsDictV2 {
    return this.store.notes;
  }
  get schemas(): SchemaModuleDictV2 {
    return this.store.schemas;
  }

  /**
   * Does not throw error but returns it
   */
  async init(): Promise<DEngineInitRespV2> {
    try {
      const { data, error } = await this.store.init();
      const { notes, schemas } = data;
      this.updateIndex("note");
      this.updateIndex("schema");
      return {
        error,
        data: { notes, schemas },
      };
    } catch (error) {
      const { message, stack, msg, status, friendly } = error;
      let payload = { message, stack };
      return {
        error: new DendronError({ payload, msg, status, friendly }),
        data: {
          notes: {},
          schemas: {},
        },
      };
    }
  }

  async deleteNote(
    id: string,
    opts?: EngineDeleteOptsV2
  ): ReturnType<DEngineClientV2["deleteNote"]> {
    try {
      const note = this.notes[id];
      const changed = await this.store.deleteNote(id, opts);
      const noteChangeEntry = _.find(
        changed,
        (ent) => ent.note.id === id
      ) as NoteChangeEntry;
      if (noteChangeEntry.status === "delete") {
        await this.fuseEngine.removeNoteFromIndex(note);
      }
      return {
        data: changed,
        error: null,
      };
    } catch (err) {
      return {
        data: [],
        error: err,
      };
    }
  }

  async deleteSchema(
    id: string,
    opts?: EngineDeleteOptsV2
  ): Promise<DEngineDeleteSchemaRespV2> {
    try {
      const data = (await this.store.deleteSchema(
        id,
        opts
      )) as DEngineDeleteSchemaRespV2;
      // deleted schema might affect notes
      await this.updateIndex("note");
      await this.updateIndex("schema");
      return data;
      // FIXM:E not performant
      // const smod = this.schemas[id];
      // await this.fuseEngine.removeSchemaFromIndex(smod);
      // return {
      //   data: undefined,
      //   error: null,
      // };
    } catch (err) {
      return {
        data: {
          notes: {},
          schemas: {},
        },
        error: err,
      };
    }
  }

  async getNoteByPath({
    npath,
    createIfNew,
    vault,
  }: GetNoteOptsV2): Promise<RespV2<GetNotePayloadV2>> {
    const ctx = "getNoteByPath";
    this.logger.debug({ ctx, npath, createIfNew, msg: "enter" });
    const maybeNote = NoteUtilsV2.getNoteByFname(npath, this.notes, {
      vault: vault,
    });
    this.logger.debug({ ctx, maybeNote, msg: "post-query" });
    let noteNew: NotePropsV2 | undefined = maybeNote;
    let changed: NoteChangeEntry[] = [];
    let error = null;
    if ((!maybeNote || maybeNote.stub) && createIfNew) {
      this.logger.debug({ ctx, maybeNote, msg: "create-new" });
      if (maybeNote?.stub) {
        noteNew = maybeNote;
        delete noteNew.stub;
      } else {
        noteNew = NoteUtilsV2.createWithSchema({
          noteOpts: { fname: npath, vault },
          engine: this,
        });
      }
      changed = (await this.writeNote(noteNew, { newNode: true })).data;
    }
    if (!createIfNew && !maybeNote) {
      error = new DendronError({ status: "no_note_found" });
    }
    return {
      data: { note: noteNew, changed },
      error,
    };
  }

  async getConfig() {
    const cpath = DConfig.configPath(this.configRoot);
    const config = readYAML(cpath) as DendronConfig;
    return {
      error: null,
      data: config,
    };
  }

  async getSchema(id: string): Promise<RespV2<SchemaModulePropsV2>> {
    const ctx = "getSchema";
    const data = this.schemas[id];
    this.logger.info({ ctx, msg: "exit" });
    return {
      data,
      error: null,
    };
  }

  queryNotesSync({
    qs,
  }: {
    qs: string;
  }): ReturnType<DEngineClientV2["queryNotesSync"]> {
    const items = this.fuseEngine.queryNote({ qs });
    return {
      error: null,
      data: items.map((ent) => this.notes[ent.id]),
    };
  }

  async querySchema(queryString: string): Promise<SchemaQueryResp> {
    const ctx = "querySchema";

    let items: SchemaModulePropsV2[] = [];
    const results = await this.fuseEngine.querySchema({ qs: queryString });
    items = results.map((ent) => this.schemas[ent.id]);
    // if (queryString === "") {
    //   items = [this.schemas.root];
    // } else if (queryString === "*") {
    //   items = _.values(this.schemas);
    // } else {
    //   const results = this.schemaIndex.search(queryString);
    //   items = _.map(results, (resp) => this.schemas[resp.item.id]);
    // }
    this.logger.info({ ctx, msg: "exit" });
    return {
      error: null,
      data: items,
    };
  }

  async queryNotes(
    opts: QueryNotesOpts
  ): ReturnType<DEngineClientV2["queryNotes"]> {
    const ctx = "Engine:queryNotes";
    const { qs, vault, createIfNew } = opts;
    let items = await this.fuseEngine.queryNote({ qs });
    let item = this.notes[items[0].id];
    if (createIfNew) {
      let noteNew: NotePropsV2;
      if (item?.fname === qs && item?.stub) {
        noteNew = item;
        noteNew.stub = false;
      } else {
        if (_.isUndefined(vault)) {
          return {
            error: new DendronError({ msg: "no vault specified" }),
            data: null as any,
          };
        }
        noteNew = NoteUtilsV2.create({ fname: qs, vault });
      }
      await this.writeNote(noteNew, { newNode: true });
    }
    this.logger.info({ ctx, msg: "exit" });
    let notes = items.map((ent) => this.notes[ent.id]);
    if (!_.isUndefined(vault)) {
      notes = notes.filter((ent) => VaultUtils.isEqual(vault, ent.vault));
    }
    return {
      error: null,
      data: notes,
    };
  }

  // async query(
  //   queryString: string,
  //   mode: DNodeTypeV2,
  //   opts?: QueryOptsV2
  // ): Promise<EngineQueryNoteResp> {
  //   const ctx = "Engine:query";
  //   const cleanOpts = _.defaults(opts || {}, {
  //     fullNode: false,
  //     createIfNew: false,
  //     initialQuery: false,
  //     stub: false,
  //   });
  //   this.logger.info({ ctx, msg: "enter" });
  //   let items: DNodePropsV2[] = [];

  //   // ~~~ schema query
  //   if (mode === "schema") {
  //     throw Error("engine.query for schema is not supported");
  //   } else {
  //     // ~~~ note query
  //     items = await this.fuseEngine.queryNote({ qs: queryString });
  //     // if (queryString === "") {
  //     //   items = [this.notes.root];
  //     // } else if (queryString === "*") {
  //     //   items = _.values(this.notes);
  //     // } else {
  //     //   const results = this.notesIndex.search(queryString);
  //     //   items = _.map(results, (resp) => resp.item);
  //     // }
  //     if (cleanOpts.createIfNew) {
  //       let noteNew: NotePropsV2;
  //       if (items[0]?.fname === queryString && items[0]?.stub) {
  //         noteNew = items[0];
  //         noteNew.stub = false;
  //       } else {
  //         noteNew = NoteUtilsV2.create({ fname: queryString });
  //       }
  //       await this.writeNote(noteNew, { newNode: true });
  //     }
  //     if (cleanOpts.fullNode) {
  //       throw Error("not implemented");
  //     }
  //   }

  //   // ~~~ exit
  //   this.logger.info({ ctx, msg: "exit" });
  //   return {
  //     error: null,
  //     data: items,
  //   };
  // }

  async sync() {
    throw Error("sync not implemented");
    return {} as any;
  }

  async renameNote(opts: RenameNoteOptsV2): Promise<RespV2<RenameNotePayload>> {
    const resp = await this.store.renameNote(opts);
    return {
      error: null,
      data: resp,
    };
  }

  async updateNote(
    note: NotePropsV2,
    opts?: EngineUpdateNodesOptsV2
  ): Promise<void> {
    const out = this.store.updateNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async updateIndex(mode: DNodeTypeV2) {
    if (mode === "schema") {
      this.fuseEngine.updateSchemaIndex(this.schemas);
    } else {
      this.fuseEngine.updateNotesIndex(this.notes);
    }
  }

  async updateSchema(schemaModule: SchemaModulePropsV2) {
    const out = await this.store.updateSchema(schemaModule);
    await this.updateIndex("schema");
    return out;
  }

  async writeConfig(
    opts: ConfigWriteOpts
  ): ReturnType<DEngineV2["writeConfig"]> {
    const { configRoot } = this;
    const cpath = DConfig.configPath(configRoot);
    writeYAML(cpath, opts.config);
    return {
      error: null,
    };
  }

  async writeNote(
    note: NotePropsV2,
    opts?: EngineWriteOptsV2
  ): Promise<WriteNoteResp> {
    const out = await this.store.writeNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async writeSchema(schema: SchemaModulePropsV2) {
    return this.store.writeSchema(schema);
  }
}
