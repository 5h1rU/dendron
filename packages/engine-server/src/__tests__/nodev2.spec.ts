import { DEngineV2, NoteUtilsV2, SchemaUtilsV2 } from "@dendronhq/common-all";
import { createLogger, note2File } from "@dendronhq/common-server";
import {
  EngineTestUtilsV2,
  NodeTestUtilsV2,
} from "@dendronhq/common-test-utils";
import fs from "fs-extra";
import path from "path";
import { FileStorageV2 } from "../drivers/file/storev2";
import { DendronEngineV2 } from "../enginev2";

let LOGGER = createLogger("enginev2.spec", "/tmp/engine-server.log");

const beforePreset = async () => {
  const vaultDir = await EngineTestUtilsV2.setupVault({
    initDirCb: async (vaultPath: string) => {
      await NodeTestUtilsV2.createSchemas({ vaultPath });
      await NodeTestUtilsV2.createNotes({ vaultPath });
      await NodeTestUtilsV2.createNoteProps({ vaultPath, rootName: "foo" });
      await NodeTestUtilsV2.createSchemaModuleOpts({
        vaultDir: vaultPath,
        rootName: "foo",
      });
    },
  });
  const engine = new DendronEngineV2({
    vaults: [vaultDir],
    forceNew: true,
    store: new FileStorageV2({ vaults: [vaultDir], logger: LOGGER }),
    mode: "fuzzy",
    logger: LOGGER,
  });
  return { vaultDir, engine };
};

describe("note", () => {
  describe("serialize", () => {
    test("basic", () => {
      const note = NoteUtilsV2.create({
        id: "foo",
        fname: "foo",
        created: "1",
        updated: "1",
      });
      const serialized = NoteUtilsV2.serialize(note);
      expect(serialized).toMatchSnapshot();
      expect(serialized.indexOf("stub") >= 0).toBeFalsy();
    });

    test("with children", () => {
      const note = NoteUtilsV2.create({
        id: "foo",
        fname: "foo",
        created: "1",
        updated: "1",
        children: ["ch1", "ch2"],
      });
      const serialized = NoteUtilsV2.serialize(note);
      expect(serialized).toMatchSnapshot();
    });

    test("with parent", () => {
      const note = NoteUtilsV2.create({
        id: "foo",
        fname: "foo",
        created: "1",
        updated: "1",
        parent: "root",
      });
      const serialized = NoteUtilsV2.serialize(note);
      expect(serialized).toMatchSnapshot();
    });

    test("with custom", () => {
      const note = NoteUtilsV2.create({
        id: "foo",
        fname: "foo",
        created: "1",
        updated: "1",
        custom: {
          bond: 42,
        },
      });
      const serialized = NoteUtilsV2.serialize(note);
      expect(serialized).toMatchSnapshot();
      // should be at beginning of line
      expect(serialized.match(/^bond/gm)).toBeTruthy();
    });

    test("with hierarchy", () => {
      const note = NoteUtilsV2.create({
        id: "foo",
        fname: "foo",
        created: "1",
        updated: "1",
        children: ["ch1", "ch2"],
        parent: "root",
      });
      const serialized = NoteUtilsV2.serialize(note, { writeHierarchy: true });
      expect(serialized).toMatchSnapshot();
      expect(serialized.match(/^parent: root/gm)).toBeTruthy();
      expect(serialized.match(/ch1/gm)).toBeTruthy();
      expect(serialized.match(/ch2/gm)).toBeTruthy();
    });

    test("with hierarchy and null parent", () => {
      const note = NoteUtilsV2.create({
        id: "foo",
        fname: "foo",
        created: "1",
        updated: "1",
        children: ["ch1", "ch2"],
        parent: null,
      });
      const serialized = NoteUtilsV2.serialize(note, { writeHierarchy: true });
      expect(serialized).toMatchSnapshot();
      expect(serialized.match(/^parent: null/gm)).toBeTruthy();
      expect(serialized.match(/ch1/gm)).toBeTruthy();
      expect(serialized.match(/ch2/gm)).toBeTruthy();
    });
  });
});

describe("matchPath", () => {
  let engine: DEngineV2;

  beforeEach(async () => {
    ({ engine } = await beforePreset());
  });

  it("match path on domain, reg", async () => {
    await engine.init();
    const resp = SchemaUtilsV2.matchPath({
      notePath: "foo",
      schemaModDict: engine.schemas,
    });
    expect(resp?.schema.id).toEqual("foo");
  });

  it("match path on domain as namespace", async () => {
    await engine.init();
    const schema = SchemaUtilsV2.createModuleProps({ fname: "bond" });
    schema.schemas[schema.root.id].data.namespace = true;
    await engine.init();
    await engine.updateSchema(schema);
    const resp = SchemaUtilsV2.matchPath({
      notePath: "bond",
      schemaModDict: engine.schemas,
    });
    expect(resp?.schema.id).toEqual("bond");
  });

  it("match path on domain as namespace, child", async () => {
    await engine.init();
    const schema = SchemaUtilsV2.createModuleProps({ fname: "bond" });
    schema.schemas[schema.root.id].data.namespace = true;
    await engine.init();
    await engine.updateSchema(schema);
    const resp = SchemaUtilsV2.matchPath({
      notePath: "bond.foo",
      schemaModDict: engine.schemas,
    });
    expect(resp?.schema.id).toEqual("bond");
  });
});

describe("matchDomain", () => {
  let vaultDir: string;
  let engine: DEngineV2;

  beforeEach(async () => {
    ({ vaultDir, engine } = await beforePreset());
  });

  it("match path on domain, reg", async () => {
    await engine.init();
    const schema = engine.notes["foo"].schema;
    expect(schema).toMatchSnapshot();
    expect(schema).toEqual({ moduleId: "foo", schemaId: "foo" });
  });

  it("match path on domain as namespace", async () => {
    await NodeTestUtilsV2.createSchemaModuleOpts({
      vaultDir,
      rootName: "bond",
      rootOpts: {
        data: { namespace: true },
      },
    });
    await NodeTestUtilsV2.createNoteProps({
      vaultPath: vaultDir,
      rootName: "bond",
    });
    fs.removeSync(path.join(vaultDir, "foo.schema.yml"));
    await engine.init();
    const schema = engine.notes["bond"].schema;
    expect(schema).toMatchSnapshot();
    expect(schema).toEqual({ moduleId: "bond", schemaId: "bond" });
  });

  it("match path on domain as namespace", async () => {
    await NodeTestUtilsV2.createSchemaModuleOpts({
      vaultDir,
      rootName: "bond",
      rootOpts: {
        data: { namespace: true },
      },
    });
    const rootName = "bond";
    const ch1 = NoteUtilsV2.create({
      fname: `${rootName}.ch1`,
      id: `${rootName}.ch1`,
      created: "1",
      updated: "1",
    });
    await note2File(ch1, vaultDir);
    fs.removeSync(path.join(vaultDir, "foo.schema.yml"));
    await engine.init();
    const schema = engine.notes["bond.ch1"].schema;
    expect(schema).toMatchSnapshot();
    expect(schema).toEqual({ moduleId: "bond", schemaId: "bond" });
  });
});
