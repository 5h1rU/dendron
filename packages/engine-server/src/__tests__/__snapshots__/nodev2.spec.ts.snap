// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchDomain match path on domain as namespace 1`] = `
Object {
  "moduleId": "bond",
  "schemaId": "bond",
}
`;

exports[`matchDomain match path on domain as namespace 2`] = `
Object {
  "moduleId": "bond",
  "schemaId": "bond",
}
`;

exports[`matchDomain match path on domain, reg 1`] = `
Object {
  "moduleId": "foo",
  "schemaId": "foo",
}
`;

exports[`matchPath match path on domain as namespace 1`] = `
Object {
  "namespace": true,
  "notePath": "bond",
  "schema": Object {
    "body": "",
    "children": Array [],
    "created": "1",
    "data": Object {
      "namespace": true,
    },
    "desc": "",
    "fname": "bond",
    "id": "bond",
    "links": Array [],
    "parent": "root",
    "title": "bond",
    "type": "schema",
    "updated": "1",
  },
  "schemaModule": Object {
    "fname": "bond",
    "imports": Array [],
    "root": Object {
      "body": "",
      "children": Array [],
      "created": "1",
      "data": Object {
        "namespace": true,
      },
      "desc": "",
      "fname": "bond",
      "id": "bond",
      "links": Array [],
      "parent": "root",
      "title": "bond",
      "type": "schema",
      "updated": "1",
    },
    "schemas": Object {
      "bond": Object {
        "body": "",
        "children": Array [],
        "created": "1",
        "data": Object {
          "namespace": true,
        },
        "desc": "",
        "fname": "bond",
        "id": "bond",
        "links": Array [],
        "parent": "root",
        "title": "bond",
        "type": "schema",
        "updated": "1",
      },
    },
    "version": 1,
  },
}
`;

exports[`matchPath match path on domain as namespace, child 1`] = `
Object {
  "namespace": false,
  "notePath": "bond.foo",
  "schema": Object {
    "body": "",
    "children": Array [],
    "created": "1",
    "data": Object {
      "namespace": true,
    },
    "desc": "",
    "fname": "bond",
    "id": "bond",
    "links": Array [],
    "parent": "root",
    "title": "bond",
    "type": "schema",
    "updated": "1",
  },
  "schemaModule": Object {
    "fname": "bond",
    "imports": Array [],
    "root": Object {
      "body": "",
      "children": Array [],
      "created": "1",
      "data": Object {
        "namespace": true,
      },
      "desc": "",
      "fname": "bond",
      "id": "bond",
      "links": Array [],
      "parent": "root",
      "title": "bond",
      "type": "schema",
      "updated": "1",
    },
    "schemas": Object {
      "bond": Object {
        "body": "",
        "children": Array [],
        "created": "1",
        "data": Object {
          "namespace": true,
        },
        "desc": "",
        "fname": "bond",
        "id": "bond",
        "links": Array [],
        "parent": "root",
        "title": "bond",
        "type": "schema",
        "updated": "1",
      },
    },
    "version": 1,
  },
}
`;

exports[`note serialize basic 1`] = `
"---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

"
`;

exports[`note serialize with children 1`] = `
"---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

"
`;

exports[`note serialize with custom 1`] = `
"---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
bond: 42
---

"
`;

exports[`note serialize with hierarchy 1`] = `
"---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
parent: root
children:
  - ch1
  - ch2
---

"
`;

exports[`note serialize with hierarchy and null parent 1`] = `
"---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
parent: null
children:
  - ch1
  - ch2
---

"
`;

exports[`note serialize with parent 1`] = `
"---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

"
`;
